Clean up `sg_params_checker` so it can be run on `data-multi-subject` without crashing

joshuacwnewton opened the pull request:
## Description

This PR has 3 parts:

- The first 16 commits are refactoring commits to make the script clearer and more concise. They clean up the code style of the script, while also adding documentation to make certain aspects (`BIDSLayout`, `BIDSLayoutIndexer`) easier to understand.
- The second to last commit, dc73e639e714ce4c64286b1b013d92386a797954, fixes #264.
- The last commit, 4adaeb7f09dfe5e66a2cb940a25ceccf1a4a88a4, adds some TODOs for discussion in this PR. 

Running the script on `data-multi-subject` results in 142 warnings:

<details>

```
WARNING: sub-amu01_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-amu01_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-amu02_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-amu02_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-amu03_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-amu03_T2w.nii.gz: Incorrect FlipAngle: FA=135 instead of 120.
WARNING: sub-amu04_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-amu04_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-amu05_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-amu05_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-beijingVerio01_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-beijingVerio01_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-beijingVerio02_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-beijingVerio02_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-beijingVerio03_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-beijingVerio03_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-beijingVerio04_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-beijingVerio04_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-nottwil01_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil01_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil01_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil01_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil01_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil01_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil02_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil03_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil04_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil05_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-nottwil06_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-strasbourg01_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-strasbourg01_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-strasbourg02_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-strasbourg02_T2w.nii.gz: Incorrect FlipAngle: FA=180 instead of 120.
WARNING: sub-ucdavis01_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis01_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis01_flip-2_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis01_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis01_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis01_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_flip-2_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis02_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis03_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis03_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis03_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis03_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis03_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_flip-2_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis04_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis05_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis05_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis05_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis05_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis05_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis06_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis06_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis06_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis06_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis06_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis07_flip-1_mt-off_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis07_flip-1_mt-on_MTS.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis07_T1w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis07_T2star.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucdavis07_T2w.nii.gz: Model 'MAGNETOM Vida' not present in list of known models for manufacturer 'Siemens'. Cannot check parameters.
WARNING: sub-ucl01_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl01_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl01_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl01_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl01_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl01_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl02_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl03_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl04_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl05_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_flip-1_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_flip-1_mt-on_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_flip-2_mt-off_MTS.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_T1w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_T2star.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-ucl06_T2w.nii.gz: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.
WARNING: sub-vallHebron01_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron01_T2w.nii.gz: Incorrect FlipAngle: FA=170 instead of 180.
WARNING: sub-vallHebron02_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron03_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron04_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron04_T2w.nii.gz: Incorrect FlipAngle: FA=165 instead of 180.
WARNING: sub-vallHebron05_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron06_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron07_T2w.nii.gz: Incorrect RepetitionTime: TR=2 instead of 1.5 +/- 0.1.
WARNING: sub-vallHebron07_T2w.nii.gz: Incorrect FlipAngle: FA=150 instead of 180.
```
</details>

## Reviewing this PR.

I hope the refactoring commits aren't too much of a bother; I've tried my best to keep them clear and atomic. So, I'd recommend using the "Commits" view to review each change in isolation.

##  Related issues

Fixes #264.

joshuacwnewton commented:
Failures all across the board here.

- Linting issues --> Should be fixed-ish by https://github.com/spine-generic/spine-generic/pull/263
- Test errors: I'm seeing the following traceback across all Python versions:
    ```
     Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.7.16/x64/bin/sg_generate_figure", line 11, in <module>
        load_entry_point('spinegeneric', 'console_scripts', 'sg_generate_figure')()
      File "/opt/hostedtoolcache/Python/3.7.16/x64/lib/python3.7/site-packages/pkg_resources/__init__.py", line 490, in load_entry_point
        return get_distribution(dist).load_entry_point(group, name)
      File "/opt/hostedtoolcache/Python/3.7.16/x64/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2862, in load_entry_point
        return ep.load()
      File "/opt/hostedtoolcache/Python/3.7.16/x64/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2462, in load
        return self.resolve()
      File "/opt/hostedtoolcache/Python/3.7.16/x64/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2468, in resolve
        module = __import__(self.module_name, fromlist=['__name__'], level=0)
      File "/home/runner/work/spine-generic/spine-generic/spinegeneric/cli/generate_figure.py", line 32, in <module>
        from sklearn.linear_model import LinearRegression
    ModuleNotFoundError: No module named 'sklearn'
    ```

`sklearn` is in spine-generic's `requirements.txt`:

https://github.com/spine-generic/spine-generic/blob/3aa69465063d823088b2c819d3ad9b81725b2fc8/requirements.txt#L7

But looking at the install logs, what actually gets installed is:

```
Collecting sklearn
  Downloading sklearn-0.0.post1.tar.gz (3.6 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
```

This... doesn't feel correct. My hunch is that we should instead be specifying [`scikit-learn`](https://pypi.org/project/scikit-learn/) in our requirements.txt? I'm a tad confused as to how this flew under the radar...

mguaypaq commented:
Re: scikit-learn, looking at [sklearn on PyPI](https://pypi.org/project/sklearn/), it looks like `sklearn` is a deprecated synonym for the package name, so that might explain things. In fact, I notice that they have brownout periods; I wonder if that's been causing intermittent failures in some of our projects? I definitely noticed intermittent errors on GitHub Actions, though I don't remember which repos.

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+    # Fetch a list of `BIDSImageFile` objects from the layout that meet the requirements below
+    query = layout.get(suffix=["T1w", "T2w", "T2star", "MTS"], extension="nii.gz")
 
+    # Fetch acquisition parameters from various vendors (Siemens, GE, Phillips) and MRI models
```
```suggestion
    # Fetch acquisition parameters for various vendors (Siemens, GE, Phillips) and MRI models
```

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+        # Check that the json sidecar has the correct keys and values
+        if "Manufacturer" not in item.get_metadata():
+            logging.warning(f" {item.filename}: Missing 'Manufacturer' key in json sidecar; Cannot check parameters.")
+            continue
```
Thanks for improving the code flow and reducing the nesting!

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+        # 'MToff_MTS', 'T1w_MTS' etc. will be used. So, we need to parse the type of MTS from the filename,
+        # then convert it to the specific names expected by the 'manufacturer params' dictionary.
+        if Contrast == "MTS":
+            # TODO: The manufacturer param dicts specifically contains the key "T1w_MTS", but I can't
+            #       find a single file in `data-multi-subject` that is MTS + T1w. Instead, all I see are
+            #       'mt-off' and 'mt-on'. So, this new method for parsing filenames passes for
+            #       data-multi-subject, but may fail for "T1w_MTS" data, if such data even exists?
```
You're not finding `T1w_MTS` files because they've already been renamed, in https://github.com/spine-generic/data-multi-subject/commit/d9a155bad5286615697f20870b094f2048e90f3d. They now contain `flip-2` in the file name instead.

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+            #       data-multi-subject, but may fail for "T1w_MTS" data, if such data even exists?
+            try:
+                # Try new method for renamed, BIDS-compliant 'data-multi-subject'
+                MTS_acq = item.filename.split('_')[-2]
```
This isn't exactly the right way to do it, but this PR is already such an improvement in the existing functionality that I'm tempted to just fix it in a follow-up PR.

joshuacwnewton replied:
I've tweaked the approach to account for https://github.com/spine-generic/spine-generic/pull/265#discussion_r1164653809 in https://github.com/spine-generic/spine-generic/pull/265/commits/5ebad16740eee1d86ada73b657c86cf99f57af1e.

Is this closer to the correct approach?

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+        # Fetch the names of each available parameter for the given manufacturer + model
+        keys_contrast = data[Manufacturer][ManufacturersModelName][str(Contrast)].keys()
+
+        # Validate repetition time against manufacturer's specifications
```
The specs are actually _about_ manufacturers, but _by_ the spine-generic protocol.
```suggestion
        # Validate repetition time against spine-generic specifications
```

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+        RepetitionTime = item.get_metadata()["RepetitionTime"]
+        if "RepetitionTime" in keys_contrast:
+            ExpectedRT = data[Manufacturer][ManufacturersModelName][str(Contrast)]["RepetitionTime"]
+            # TODO: We check against at threshold here, rather than FA, which checks for exactness. Do we want this?
```
I think RepetitionTime is really a floating point number, in which case fuzzy checking is the right thing to do.

In contrast, FlipAngle values seem to be integers, specifically in the list [9, 15, 30, 90, 120, 180], which look like discrete values, in which case exact equality is correct.

The thing that worries me with this fuzzy equality check is that it's one-sided! I would really expect `abs(...) > 0.1`, but I'm not sure what the original intent is.

joshuacwnewton replied:
> The thing that worries me with this fuzzy equality check is that it's one-sided! I would really expect `abs(...) > 0.1`, but I'm not sure what the original intent is.

Very very good point. I'll update the TODOs to mention this instead.

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
+                logging.warning(f" {item.filename}: Incorrect RepetitionTime: "
+                                f"TR={RepetitionTime} instead of {ExpectedRT} +/- 0.1.")
+
+        # Validate echo time against manufacturer's specifications
```
Ditto.
```suggestion
        # Validate echo time against spine-generic specifications
```

mguaypaq reviewed (approved):
Thanks for the cleanup!

This is already a big improvement, and I don't think it changes any of the existing functionality, so I'm tempted to merge it as a base for further improvements.

Alternatively, I could just add a few commits on this branch if you prefer.

joshuacwnewton commented on requirements.txt:
```diff
@@ -4,7 +4,7 @@ nibabel
 scipy
 tqdm
-sklearn
+scikit-learn
```
I can also commit this change directly to master and remove it from this PR, just to keep things tidy.

joshuacwnewton commented:
I think I've addressed most of your feedback! If you feel it looks good (and that further concerns can be tweaked in follow-up PRs), then feel free to merge! :tada:

mguaypaq commented on spinegeneric/cli/params_checker.py:
```diff
@@ -54,113 +53,101 @@ def main():
             data_path,
-            indexer=BIDSLayoutIndexer(config_filename=path_sg_layout_config),
+            # BIDSLayoutIndexer is a class that indexes files based on pattern-matching defined in the config.
+            # By default, BIDS has its own config. But, SG specifies its own custom config instead. (Why?)
```
I tracked it down:
- The `bids_specs.json` file was introduced in #198, specifically because at the time the default config file didn't support the `MTS` suffix in filenames. Thankfully the corresponding issue #190 links to the then-current upstream config file.
- Comparing `bids_specs.json` with the historical upstream config, only two lines are different, inside `default_path_patterns`, and they just add `MTS` as valid suffixes.
- Since [pybids#840](https://github.com/bids-standard/pybids/pull/840) from last year, upstream now supports the `MTS` suffix by default, but with a difference: it looks like `flip` and `mt` are mandatory entities in the filename when the suffix is `MTS`.

I think the upshot is that we can get rid of the custom config, and of the old handling of `MTS` (which relied on the `acq` entity, instead of `flip` and `mt`). I'll write a separate PR for that.

joshuacwnewton replied:
Wonderful sleuthing!! :hearts:
